### PR TITLE
Create course groups in JIRA automatically based on the course short name

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I updated multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] ~Client: I added multiple integration tests (Jest) related to the features~
- [x] ~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~
- [x] ~Client: I documented the TypeScript code using JSDoc style.~
- [x] ~Client: I added multiple screenshots/screencasts of my UI changes~
- [x] ~Client: I translated all the newly inserted strings into German and English~

### Motivation and Context
Simplify course administration in JIRA
Addresses #1221 

### Description
Improve error handling for courses and programming exercises
Check that short names are always unique
In case of an error, automatically scroll up in the client

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a course
4. Edit a course
5. Delete a course

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 1]
```diff
@@ -620,7 +619,20 @@ private boolean userHasPermission(Course course, User user) {
...
+        // only delete the groups if they have been created by Artemis before
+        if (course.getStudentGroupName().equals("artemis-" + course.getShortName() + "-students")) {
+            artemisAuthenticationProvider.deleteGroup(course.getStudentGroupName());
+        }
+        if (course.getStudentGroupName().equals("artemis-" + course.getShortName() + "-tutors")) {
+            artemisAuthenticationProvider.deleteGroup(course.getTeachingAssistantGroupName());
+        }
+        if (course.getStudentGroupName().equals("artemis-" + course.getShortName() + "-instructors")) {
+            artemisAuthenticationProvider.deleteGroup(course.getInstructorGroupName());
+        }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]:
```suggestion
        if (course.getTeachingAssistantGroupName().equals("artemis-" + course.getShortName() + "-tutors")) {
            artemisAuthenticationProvider.deleteGroup(course.getTeachingAssistantGroupName());
        }
        if (course.getInstructorGroupName().equals("artemis-" + course.getShortName() + "-instructors")) {
            artemisAuthenticationProvider.deleteGroup(course.getInstructorGroupName());
        }
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 2]
```diff
@@ -141,22 +136,48 @@ public CourseResource(Environment env, UserService userService, CourseService co
...
+        if (course.getInstructorGroupName() != null) {
+            throw new BadRequestAlertException("The instructor group name must be null when creating a course", ENTITY_NAME, "instructorGroupNameCannotBeSet");
+        }
+
         validateComplaintsConfig(course);
+
+        course.setStudentGroupName("artemis-" + course.getShortName() + "-students");
+        course.setTeachingAssistantGroupName("artemis-" + course.getShortName() + "-tutors");
+        course.setInstructorGroupName("artemis-" + course.getShortName() + "-instructors");
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]: I am wondering if we should create the group names in the Course Object, as we use them multiple times.

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]: good point. I refactored it

**Comment by krusche** [comment]
@jpbernius thanks for the feedback. I addressed it. Now, it should be fine :-)

**Pull request merged**